Reduce all tensors to their metadata in AOTAutogradCache; add tests

jamesjwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130252
* #129384
* __->__ #128583



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

This PR makes it so that all tensors are reduced to their metadata in AOTAutogradCache. Because dynamo always embeds constant tensors into the FXgraph directly, there's no risk of a constant tensor whose values are semantically important being lost here. AOTAutograd itself may take a constant tensor and set it as an attribute on an FXGraph for inductor, but Dynamo never does this.

One other thing that this diff does is add `[pickler.fast](https://docs.python.org/3/library/pickle.html#pickle.Pickler.fast)` to our pickling algorithm for cache key generation. Pickle will often memoize/intern strings when pickling, leading to false cache misses due to inconsistent memoization. Turning on pickler.fast removes this behavior. 

Technically `fast` is a "deprecated" feature according to python docs. But it's still supported in py3.8-3.12, and if it ever is removed, the only downside will just be a few more cache misses, so I think it's worth just adding here (and removing later as needed)

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Code review

[ghstack-poisoned]

oulgen commented on torch/_inductor/codecache.py:
```diff
@@ -526,6 +526,8 @@ def dumps(cls, obj) -> bytes:
         with io.BytesIO() as stream:
             pickler = cls(stream)
+            # TODO: pickler.fast is technically deprecated. Will this work on new python versions?
+            pickler.fast = True  # Run with pickler.fast so it doesn't intern strings, making the hash result more predictable
```
this feels like a time bomb

jamesjwu replied:
Here's my reasoning why it isn't, but feel free to disagree:
- The way this breaks is if we upgrade python version, right? As of Python 3.12 this feature is supported, and when we upgrade python version, it's not like we'll instantly move to it without a PR to fix existing issues. So Python would need to 1) fully remove the feature 2) cause it to crash instead of, say, silently doing nothing 3) release it in a new version of python. 
- Even after all of that, because it's python, even *if* they remove the fast feature from picklers, setting "fast" on an object in python won't crash: it might do nothing, which would just lead to the behavior of not adding it. 
- If we want to be extra safe, we could add a version check here as well.

oulgen replied:
I'm more worried about silent failure and the next steps. We need to be able to detect that we have cache misses which will be hard because it we will still get hits since this is not a common occurence. Secondly, we don't actually know how to fix it when this happens, so this will be on the critical path for say python 3.13 upgrade. I'd be more okay with it I think if we knew how to fix it but choose not to do it yet.

jamesjwu replied:
One other thing: this makes it so we can't pickle cyclic objects like:
```
class MyNode:
    def __init__(self):
        self.name = self
```

However, on trying to pickle one of these, we'll get a safe Pickle ValueError like so:
```
*** ValueError: fast mode: can't pickle cyclic objects including object type dict at 0x7f7079dd8980
```
which will naturally lead to a BypassGraphCache and a cache miss. So if we ever do create cyclic objects (which we don't today), we will still fail fast.

jamesjwu replied:
That said, my opinion on this isn't super strong because I feel like in real life no one will ever hit this cache miss too seriously (since it looks like pickle will be deterministic anyway on process start). But it does feel like a free win to me, where the only downside is we might have to remove the win later.

jamesjwu replied:
To elaborate on why I think it's a "free win": checking https://docs.python.org/es/dev/whatsnew/3.13.html#new-deprecations, it doesn't look like Pickler.fast is being removed, up to and including Python 3.16. Second of all, I'm not sure if it's really in the critical path: if, say, Python 3.17 removes Pickler.fast, we just stop doing this optimization and cache miss more, right? At worst, we've reverted back to the existing behavior without it.

masnesral replied:
My 2c. I don't like it because it feels like a hack, but I think it's much less of a hack than the one I already added here: https://github.com/pytorch/pytorch/blob/c012013aa6c1dc84b1f566dadfce3d7c5217b860/torch/_inductor/codecache.py#L432-L440
So I think I I'd prefer to remove my hack in favor of your hack.

> this will be on the critical path for say python 3.13 upgrade

I wonder if there's precedent for an assertion on the version just so that we can get a "callback" when someone starts work on support for the next Python version? If it "just works", we can bump the version. If not, then I presume support for any new version is a kind of a slog and we can discuss alternatives in parallel with the slog? But I can also see how that'd be obnoxious.

> we'll get a safe Pickle ValueError like so:

Not specific to this specific exception, but: what do you guys think about introducing some kind of "strict" mode that we enable for CI (but not prod) that hard fails for unexpected failures like this, rather than quietly bypassing? It would be nice to know immediately if anyone makes a change that introduces cycles, for example.

jamesjwu replied:
I actually already added that  for AOTAutogradCache in https://github.com/pytorch/pytorch/pull/128222, so feel free to add one for strict FXGraphCache as well. I use the strict setting on that specific unit test suite. If you do add one for FXGraphCache, let me know/please also turn it on in TestAOTAutogradWithCache.

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update on "Reduce all tensors to their metadata in AOTAutogradCache; add tests"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

This PR makes it so that all tensors are reduced to their metadata in AOTAutogradCache. Because dynamo always embeds constant tensors into the FXgraph directly, there's no risk of a constant tensor whose values are semantically important being lost here. AOTAutograd itself may take a constant tensor and set it as an attribute on an FXGraph for inductor, but Dynamo never does this.

One other thing that this diff does is add `[pickler.fast](https://docs.python.org/3/library/pickle.html#pickle.Pickler.fast)` to our pickling algorithm for cache key generation. Pickle will often memoize/intern strings when pickling, leading to false cache misses due to inconsistent memoization. Turning on pickler.fast removes this behavior. 

Technically `fast` is a "deprecated" feature according to python docs. But it's still supported in py3.8-3.12, and if it ever is removed, the only downside will just be a few more cache misses, so I think it's worth just adding here (and removing later as needed)

[ghstack-poisoned]

jamesjwu committed:
Update on "Reduce all tensors to their metadata in AOTAutogradCache; add tests"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

This PR makes it so that all tensors are reduced to their metadata in AOTAutogradCache. Because dynamo always embeds constant tensors into the FXgraph directly, there's no risk of a constant tensor whose values are semantically important being lost here. AOTAutograd itself may take a constant tensor and set it as an attribute on an FXGraph for inductor, but Dynamo never does this.

One other thing that this diff does is add `[pickler.fast](https://docs.python.org/3/library/pickle.html#pickle.Pickler.fast)` to our pickling algorithm for cache key generation. Pickle will often memoize/intern strings when pickling, leading to false cache misses due to inconsistent memoization. Turning on pickler.fast removes this behavior. 

Technically `fast` is a "deprecated" feature according to python docs. But it's still supported in py3.8-3.12, and if it ever is removed, the only downside will just be a few more cache misses, so I think it's worth just adding here (and removing later as needed)

[ghstack-poisoned]

jamesjwu committed:
One more rebase

[ghstack-poisoned]

jamesjwu committed:
Code review: ignore private namespaced torch functions that aren't builtins

[ghstack-poisoned]

jamesjwu committed:
Add note and comment to check_node_safe docstring

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Remove torch.* functions from safe list

[ghstack-poisoned]

jamesjwu committed:
Rebase

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Update

[ghstack-poisoned]

jamesjwu committed:
Rebase

[ghstack-poisoned]

jamesjwu commented:
@bdhirsh I think this is ready for review now

jamesjwu commented:
@pytorchbot merge -i
